packages/rolldown/src/utils/compose-js-plugins.ts

```diff
@@ -0,0 +1,276 @@
+    [K in keyof Plugin]?: NonNullable<Plugin[K]>[]
+  } = {}
+
+  plugins.forEach((plugin, index) => {
+    const pluginName = plugin.name || `Anonymous(index: ${index})`
+    names.push(pluginName)
+    ;(Object.keys(plugin) as (keyof typeof plugin)[]).forEach((pluginProp) => {
+      switch (pluginProp) {
+        case 'name':
+          break
+        case 'buildStart': {
+          const handlers = batchedHooks.buildStart ?? []
+          batchedHooks.buildStart = handlers
+          if (plugin.buildStart) {
+            handlers.push(plugin.buildStart)
+          }
+          break
+        }
+        case 'load': {
+          const handlers = batchedHooks.load ?? []
+          batchedHooks.load = handlers
+          if (plugin.load) {
+            handlers.push(plugin.load)
+          }
+          break
+        }
+        case 'transform': {
+          const handlers = batchedHooks.transform ?? []
+          batchedHooks.transform = handlers
+          if (plugin.transform) {
+            handlers.push(plugin.transform)
+          }
+          break
+        }
+        case 'resolveId': {
+          const handlers = batchedHooks.resolveId ?? []
+          batchedHooks.resolveId = handlers
+          if (plugin.resolveId) {
+            handlers.push(plugin.resolveId)
+          }
+          break
+        }
+        case 'buildEnd': {
+          const handlers = batchedHooks.buildEnd ?? []
+          batchedHooks.buildEnd = handlers
+          if (plugin.buildEnd) {
+            handlers.push(plugin.buildEnd)
+          }
+          break
+        }
+        case 'renderChunk': {
+          const handlers = batchedHooks.renderChunk ?? []
+          batchedHooks.renderChunk = handlers
+          if (plugin.renderChunk) {
+            handlers.push(plugin.renderChunk)
+          }
+          break
+        }
+        case 'augmentChunkHash':
+        case 'banner':
+        case 'generateBundle':
+        case 'moduleParsed':
+        case 'onLog':
+        case 'options':
+        case 'outputOptions':
+        case 'renderError':
+        case 'renderStart':
+        case 'resolveDynamicImport':
+        case 'writeBundle': {
+          throw new Error(
+            `Failed to compose js plugins. Plugin ${pluginName} has an unsupported hook: ${pluginProp}`,
+          )
+          break
+        }
+        default: {
+          // All known hooks should be handled above. We allow plugin to have unknown properties and we just ignore them.
+          const _executiveCheck: never = pluginProp
+        }
+      }
+    })
+  })
+
+  const composed: Plugin = {
+    name: `Composed(${names.join(', ')})`,
+  }
+
+  ;(Object.keys(batchedHooks) as (keyof typeof batchedHooks)[]).forEach(
+    (hookName) => {
+      switch (hookName) {
+        case 'buildStart': {
+          if (batchedHooks.buildStart) {
+            const batchedHandlers = batchedHooks.buildStart
+            composed.buildStart = async function (options) {
```

Comment:
What if the user expects the `buildStart` hook to execute linearly? This may generate different behavior

Reply:
This hook is intend to be async parallel: 
![image](https://github.com/user-attachments/assets/8635f6d1-d722-42eb-9b6a-b5d4173b0416)
.

Reply:
got it.